### PR TITLE
added `inheritFrom`

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
It basically says "if a property is not defined in this widget, look at that other widget instead".

The simple form looks like this:

`"inheritFrom": "otherWidgetID"`

You can also use an object to specify which properties to inherit:

```
  "inheritFrom": {
    "otherWidgetID": [ "image", "svgReplaces" ]
  }
```

And you can even provide multiple widgets (and use `"*"` for the simple "inherit all"):

```
  "inheritFrom": {
    "otherWidgetID": [ "image", "svgReplaces" ],
    "thirdWidgetID": "*"
  }
```

Please note that some very basic properties can not be inherited: `id`, `type`, `deck` and `cardType`. Also `inheritFrom` has to be defined directly in the widget state - so not through `cardDefaults` or similar constructs.

A real-life example can be found in http://212.47.248.129:3528/Bulls.

See #479.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3528/pr-test (or any other room on that server)

**Comment by bjalder26** [comment]
I added some widgets, and it appears basic widgets are inheriting text when they are set to inherit something else.  Also, timers are inheriting milliseconds, when they are set to inherit something else, and spinners are inheriting angle and value when they are set to inherit something else.  Of course, first, their property has to be deleted first (if they normally have that property) but I'm guessing they aren't supposed to inherit these things unless specifically set to do so.

**Comment by ArnoldSmith86** [comment]
Thanks for all the testing! That's really helpful!

**Comment by ArnoldSmith86** [comment]
The problem was that I didn't update the code that inherits changing properties to any widgets that `inheritFrom` the changing widget to respect the object format (only inheriting specific properties) I added later.

So all those inheritance problems were only properties that changed on-the-fly.

> If I refresh the browser, the red checker stops inheriting text (is is not supposed to inherit text), but if I go into the JSON editor, ctrl-right-click the pink checker it comes back. IN FACT, the same is true for the green checker and the X/Y slide buttons, but I can even get it to change its text from X to Y depending on which button I ctrl-right-click on.

The behaviour with the JSON editor stems from the fact that it updates _all_ the properties of the widget after each edit and not just the ones you changed.

Please test if you find further problems. :grin:

**Comment by bjalder26** [comment]
I can't find any more issues.  Everything I've tested now seems to behave as expected.  I also noticed that default values are not overridden unless they are deleted, which is again clearly expected behavior.

**Comment by ArnoldSmith86** [comment]
Marking it as draft because we discussed in the meeting that inherited properties should trigger `changeRoutine`.

I think it's a lot more work than I thought yesterday but I'll at least take a look.

**Comment by ArnoldSmith86** [comment]
Adding support for `changeRoutine` is more work than I thought.

I would rather go ahead without it for now. We can probably always add it later although it's iffy if by the time widgets already have `inheritFrom` and `changeRoutine`.

IMO it's fine not to support it.

**Comment by ArnoldSmith86** [comment]
Updated the PR description.

**Comment by RaphaelAlvez** [comment]
> Adding support for `changeRoutine` is more work than I thought.
> 
> I would rather go ahead without it for now. We can probably always add it later although it's iffy if by the time widgets already have `inheritFrom` and `changeRoutine`.
> 
> IMO it's fine not to support it.

It is ok to merge it like it is but I do think that at some point this I'll be requested again as this would be extremely useful and powerful. Btw another problem I just thought of is inheriting a changeRoutine. Have to test just to know the situation about that

**Comment by ArnoldSmith86** [comment]
> Btw another problem I just thought of is inheriting a `changeRoutine`. Have to test just to know the situation about that

It will act just like if it was defined in the widget itself. Bulls uses inherited `clickRoutine` just fine.

**Comment by bjalder26** [comment]
As far as workarounds:

Well, for now we can have a changeRoutine on the originalProperty widget that performs changes on the inheritedProperty widget (we need better names and can't use parent-child).

We could also put a changeRoutine on the inheritedProperty widget that `GET`s the originalProperty widget's properties.  However, the problem is that changes in the value in question won't trigger the changeRoutine.

You can't `GET` that originalProperty widget property without triggering the changeRoutine first, so that property can't be what triggers it.

The changeRoutine could even `GET` the property in question from the originalProperty widget, even if `inheritedFrom` changes on the fly (i.e. originalProperty widget not known ahead of time) using `FOREACH` on `${PROPERTY inheritFrom}`, but again that doesn't solve the problem of triggering the routine.

You can put a changeRoutine in the originalProperty widget that 'pings' the inheritedProperty widget every time the property changes (e.g. toggles a value like -1 to 1).  However again this requires you know beforehand what widget will inherit what from what other widget.  

There is a very convoluted solution to that too (change routine on inheritiedProperty widget `SET`s a changeRoutine on the originalProperty widget).  However then that wouldn't even get implemented until some other change happened.

So long story short, the biggest issue seems to be triggering the changeRoutine of the inheritedProperty widget.

**Comment by ArnoldSmith86** [comment]
But: how important is this?

Isn't a solution like http://212.47.248.129:3528/changeRoutine enough?

**Comment by bjalder26** [comment]
> But: how important is this?
> 
> Isn't a solution like http://212.47.248.129:3528/changeRoutine enough?

Sorry, I wasn't commenting on how important it was.  I suspect the vast majority of uses they will know the originalProperty widget they want to use, know the inheritedProperty widget they want to use, and know the properties they want to be inherited, so I think it would be an easy workaround in the vast majority of cases.

However, it seems like you're showing in your example that you can use a CALL to make inheritedProperty widgets trigger their changeRoutine, so I guess there is a way to do that as well.

I am not suggesting this is a reason to withhold releasing the feature.

**Comment by RaphaelAlvez** [comment]
Just took a quick look into the code. Not sure I understand everything but I realized that the idea I had in mind would not work. I would suggest to display all properties and only 

Maybe I have an ugly solution for this. Instead of filtering the properties out of the JSON, just add inherited as a prefix. This would also work with inherited change routines because they would take something like x and xChangeRotuine and output inheritedx and inheritedxChangeRoutine

**Comment by ArnoldSmith86** [comment]
I don't understand. The whole point is to not have the properties in the inheriting widget.

**Comment by bjalder26** [comment]
Something weird can happen with labels in this PR.  It isn't easy to replicate, but if you open the JSON editor, select the text of a label that is inheriting its text from another label, then try to drag the text back in, then hit control and click inside the label, then hit control again and move the label it might gain the text of the label it is inheriting text from, and then it won't inherit text anymore.  I don't think this is a reason to not merge this PR.  I suspect it has something to do with the JSON editor, and shouldn't affect gameplay.  I could not get it to occur when the JSON editor was not open.

**Comment by 96LawDawg** [comment]
And that's why you are the head of testing!

**Comment by bjalder26** [comment]
> And that's why you are the head of testing!

Well, you noticed it first, I just kept at it until I found out how to replicate it (sometimes).

**Comment by rogerl50** [comment]
> > But: how important is this?
> > Isn't a solution like http://212.47.248.129:3528/changeRoutine enough?
> 
> Sorry, I wasn't commenting on how important it was. I suspect the vast majority of uses they will know the originalProperty widget they want to use, know the inheritedProperty widget they want to use, and know the properties they want to be inherited, so I think it would be an easy workaround in the vast majority of cases.
> 
> However, it seems like you're showing in your example that you can use a CALL to make inheritedProperty widgets trigger their changeRoutine, so I guess there is a way to do that as well.
> 
> I am not suggesting this is a reason to withhold releasing the feature.

@bjalder26 I've looked at the changeRoutine room above, but I can't quite figure out what the issue you were raising here was. Can you state what will or will not work after this PR is merged?

**Comment by rogerl50** [comment]
> > But: how important is this?
> > Isn't a solution like http://212.47.248.129:3528/changeRoutine enough?
> 
> Sorry, I wasn't commenting on how important it was. I suspect the vast majority of uses they will know the originalProperty widget they want to use, know the inheritedProperty widget they want to use, and know the properties they want to be inherited, so I think it would be an easy workaround in the vast majority of cases.
> 
> However, it seems like you're showing in your example that you can use a CALL to make inheritedProperty widgets trigger their changeRoutine, so I guess there is a way to do that as well.
> 
> I am not suggesting this is a reason to withhold releasing the feature.

@bjalder26 I've looked at the changeRoutine room above, but I can't quite figure out what the issue you were raising here was. Can you state what will or will not work after this PR is merged?

**Comment by bjalder26** [comment]
> > However, it seems like you're showing in your example that you can use a CALL to make inheritedProperty widgets trigger their changeRoutine, so I guess there is a way to do that as well.
> > I am not suggesting this is a reason to withhold releasing the feature.
> 
> @bjalder26 I've looked at the changeRoutine room above, but I can't quite figure out what the issue you were raising here was. Can you state what will or will not work after this PR is merged?

ArnoldSmith86 mentioned that he would like inherited properties to be able to trigger changeRoutines, and I was talking about possible workarounds.  Then he showed me that you could use `CALL` to trigger a changeRoutine as a workaround.

**Comment by rogerl50** [comment]
> ArnoldSmith86 mentioned that he would like inherited properties to be able to trigger changeRoutines, and I was talking about possible workarounds. Then he showed me that you could use `CALL` to trigger a changeRoutine as a workaround.
> 
I see. The workaround seems fine to me (but the issue should probably be documented somewhere other than this PR comment string :) )

**Comment by bjalder26** [comment]
> > ArnoldSmith86 mentioned that he would like inherited properties to be able to trigger changeRoutines, and I was talking about possible workarounds. Then he showed me that you could use `CALL` to trigger a changeRoutine as a workaround.
> 
> I see. The workaround seems fine to me (but the issue should probably be documented somewhere other than this PR comment string :) )

There will be an example in the Tutorial, but it probably wouldn't hurt for it to be mentioned in the wiki.

**Comment by ArnoldSmith86** [comment]
> Something weird can happen with labels in this PR. It isn't easy to replicate, but if you open the JSON editor, select the text of a label that is inheriting its text from another label, then try to drag the text back in, then hit control and click inside the label, then hit control again and move the label it might gain the text of the label it is inheriting text from, and then it won't inherit text anymore. I don't think this is a reason to not merge this PR. I suspect it has something to do with the JSON editor, and shouldn't affect gameplay. I could not get it to occur when the JSON editor was not open.

I couldn't reproduce this yet. What do you mean by "select the text of a label that is inheriting its text from another label, then try to drag the text back in"? Select the text inside the widget? Or the editor? If the former, is it important what's open in the editor? Then drag it out of the widget as if you wanted to copy the text into Notepad or something but then drop it back into the widget?

In which browser+OS does this happen?

**Comment by 96LawDawg** [comment]
I can answer part of that because I stumbled across it when I was in a room @bjalder26 was working on.  For me, it happened in Edge in Windows.  Not sure about the steps needed to reproduce it, but I'll try.

**Comment by 96LawDawg** [comment]
So for me, I can't get inheritFrom label to work for text at all.  No magic steps required.  See video from /Interitus test room that I just made.


https://user-images.githubusercontent.com/76912527/124327307-782b4b00-db4d-11eb-9e02-2d383223d736.mp4

**Comment by ArnoldSmith86** [comment]
> So for me, I can't get inheritFrom label to work for text at all.

Your inheriting label has its own `text` set. So it doesn't inherit it.

**Comment by 96LawDawg** [comment]
Well, that explains that.  Good to know.  But I think that did help me figure out how to reproduce the problem.  While in JSON editor, in the label that is inheriting text (correctly, without its own text property), select the text using the mouse and then press CTRL-C or CTRL-X.  A text property gets applied to the inheriting label preventing further updates.


https://user-images.githubusercontent.com/76912527/124328689-bb86b900-db4f-11eb-85ef-c6570727caf2.mp4

**Comment by ArnoldSmith86** [comment]
Please check if the latest commit helps. The `text` property of a label was always updated if a `keyup` event was fired inside the textarea. Now it only updates `text` if it is `editable: true` and if the content of the textarea actually changed.

This change is pretty much unrelated to this PR but the problem didn't matter in the past.

**Comment by 96LawDawg** [comment]
Problem fixed.

**Comment by ArnoldSmith86** [comment]
Please check if labels are still editable when they should be.

**Comment by 96LawDawg** [comment]
Yes, yes.  All that still works, also.  `editable:true` still works like it is supposed to.  Even if `editable:true` is set on the inheriting label, it still inherits so long as you don't do anything like before to cause a `keyup` or whatever.

**Comment by ArnoldSmith86** [comment]
Any other problems with this PR? It's been 27 days. :smile:

**Review by 96LawDawg — APPROVED** [review]
Approved through test (sort of).  @bjalder26 did a LOT of work on this.  He is currently on vacation.  I observed what he did and read his comments on the PR, including the fact that he thought it was ready for merge except for a minor thing that @ArnoldSmith86 found and fixed and I tested that fix.  So because of all that, I'm approving through test.

**Comment by RaphaelAlvez** [comment]
Ok I may have a solution to some problems (especially mines)

how hard would it be to introduce a originalProperty -> newProperty syntax?
I was thinking of something like this for seats
```
"inheritFrom": {
    "otherWidgetID": [ "player as text", "color as color" ]
  }
```
"player as text" would remove the necessity of code to make player labels for seats
 "color as color" is just an example that would allow to trigger changeRoutines

Properties inherited with "as otherProperty" would be copied to the widget so changeRoutines could be applied.
Note that this would not let the user overwrite them.
So why have those? this would be useful for passing changes and information inside a composed widget.

I do understand that I had this idea very late in the development of this PR and you all may want to consider this for this PR.

**Comment by ArnoldSmith86** [comment]
TBH, I think that's all easier to do with two or three routine operations.

**Review by RaphaelAlvez — APPROVED** [review]
Have used this in an staging server and it works. I agree that globalChangeRouitne can be used to do what I wanted with this PR. Maybe that can still stay open for a future update.
